x/tokenfilter/ibc_middleware.go

```diff
@@ -52,36 +54,55 @@ func (m *tokenFilterMiddleware) OnRecvPacket(
 		return m.IBCModule.OnRecvPacket(ctx, channelVersion, packet, relayer)
 	}
 
+	if len(data.Tokens) != 1 {
+		ackErr := errors.Wrapf(sdkerrors.ErrInvalidType, "expected exactly one token, got %d", len(data.Tokens))
+		return createErrorAcknowledgement(ctx, data, ackErr)
+	}
+
 	// Note, this firewall prevents receiving of any non-native token denominations as we only
 	// accept ibc tokens which were minted on behalf of this chain.
-	denom := transfertypes.ExtractDenomFromPath(data.Denom)
+	denom := data.Tokens[0].Denom
 	if receiverChainIsSource(packet.GetSourcePort(), packet.GetSourceChannel(), denom) {
 		return m.IBCModule.OnRecvPacket(ctx, channelVersion, packet, relayer)
 	}
 
-	ackErr := errors.Wrapf(sdkerrors.ErrInvalidType, "only native denom transfers accepted, got %s", data.Denom)
+	ackErr := errors.Wrapf(sdkerrors.ErrInvalidType, "only native denom transfers accepted, got %s", denom.Path())
+	return createErrorAcknowledgement(ctx, data, ackErr)
+}
 
+// receiverChainIsSource checks the first denomination prefix in the ibc transfer denom provided against the packet source port and channel.
+// If the prefix matches it means that the token was originally sent from this chain.
+// This is because OnRecvPacket prefixes the destination port and channel to the token denomination when first sent.
```

Comment:
```suggestion
// This is because ibc transfer prefixes the destination port and channel to the token denomination when minting vouchers in recv packet.
```